In [1]:
import numpy as np
import pandas as pd
pd.options.display.float_format = '{:.2f}'.format
pd.set_option("display.precision",2)
pd.set_option("chop_threshold", 0.005)

In [2]:
y = pd.Series([1,1,2,2,4],dtype=int)

In [3]:
x = pd.Series([1,2,3,4,5],dtype=int)

In [4]:
x

0    1
1    2
2    3
3    4
4    5
dtype: int32

In [5]:
y

0    1
1    1
2    2
3    2
4    4
dtype: int32

In [6]:
x.mul(y).sum()

37

# Find the b_0 and b_1  

In [13]:
def summation(x,y):
    return ((x-x.mean())  * (y- y.mean())).sum()

In [14]:
summation(x,y) /summation(x,x)

0.7

In [8]:
def sum_of_square(x,y):
    summation = x.mul(y).sum()
    return (summation - (len(x) * np.mean(x) *  np.mean(y)))

In [9]:
b_1 = round(sum_of_square(x,y) / sum_of_square(x,x) ,2)

In [10]:
b_0 = round(np.mean(y) - (b_1 * np.mean(x)),2)

In [11]:
f"Linear equation is Y={b_0}+{b_1}X"

'Linear equation is Y=-0.1+0.7X'

# Hypothesis Testing 

In [12]:
y_hat = pd.Series([(b_0+b_1*i) for i in x])

In [13]:
residual = y - y_hat

In [14]:
y_hat , residual

(0   0.60
 1   1.30
 2   2.00
 3   2.70
 4   3.40
 dtype: float64,
 0    0.40
 1   -0.30
 2    0.00
 3   -0.70
 4    0.60
 dtype: float64)

In [15]:
SS_res = (residual**2).sum()

In [16]:
dof = len(x) -2 
mean_residual_square = SS_res / dof

In [17]:
t_statistic = b_1 / (np.sqrt(mean_residual_square / sum_of_square(x,x)))

In [18]:
t_statistic

3.655630775069655

In [19]:
from scipy.stats import t

In [20]:
 2 * (1 - t.cdf(abs(t_statistic),dof))

0.03535284700251751

In [21]:
dof

3

In [23]:
p_value = t.sf(np.abs(t_statistic),dof) *2 
p_value

0.03535284700251736

In [24]:
def hypo_test(alpha,p_value):
    if(p_value < alpha):
        print("Reject the null hypothetis")
    else:
        print("Donot reject the hypothesis")

In [25]:
hypo_test(0.05,p_value)

Reject the null hypothetis


# AVONA 

In [26]:
SS_reg = b_1 **2 * sum_of_square(x,x)
SS_reg

4.8999999999999995

In [27]:
dof_reg = 1
dof_res = len(x) - 2
dof_reg ,dof_res

(1, 3)

In [28]:
MS_res = SS_res / dof_res
MS_reg = SS_reg / dof_reg
MS_reg,MS_res

(4.8999999999999995, 0.36666666666666653)

In [29]:
f_statistic = MS_reg / MS_res
f_statistic

13.363636363636367

In [30]:
from scipy.stats import f

In [31]:
p_Value = f.sf(f_statistic , dof_reg ,dof_res)
p_value

0.03535284700251736

In [32]:
hypo_test(0.05,p_value)

Reject the null hypothetis


#  Coefficient of Determination

In [51]:
SS_t = ((y - np.mean(y))**2).sum()
R_squared = SS_reg / SS_t
R_squared

0.8166666666666665

81% of varaiation is explained by the model

#  Confidence Interval

##  Chooose the Parameter

Finding the Confidence Interval for parameter(b_1)

In [54]:
alpha = 0.05
dof = len(x)-2
t.pdf(alpha/2 , dof)

0.3673994978775869

In [58]:
#to find the critical value
crit = abs(t.ppf(alpha/2 ,dof))
crit

3.1824463052842638

In [60]:
(b_1-crit* np.sqrt(MS_res/sum_of_square(x,x)) , b_1+crit* np.sqrt(MS_res/sum_of_square(x,x)))

(0.09060792766836867, 1.3093920723316312)

im 95% confident that the b_1 of the population will lie in the range interval (0.1 , 1.3)